sklearn/metrics/_pairwise_distances_reduction/_gemm_term_computer.pxd.tp

```diff
@@ -16,6 +34,12 @@ cdef class GEMMTermComputer64:
         # Buffers for the `-2 * X_c @ Y_c.T` term computed via GEMM
         vector[vector[DTYPE_t]] dist_middle_terms_chunks
 
+{{if need_upcast}}
```

Comment:
Very nit: What do you think of?

```suggestion
{{if name_suffix == "32"}}
```

It makes it more clear that 32bit requires upcasting

Reply:
That's a good remark. I think we can make clearer that the float32 implementations require upcasting whilst avoiding duplicating this logic. Also potentially in the future other dtypes might require upcast as well.

Hence, I would keep `need_upcast` but makes its definition clearer and more visible. What do you think? Furthermore, I think the naming can be improved. What about `upcast_to_float64` instead of `need_upcast`?

Reply:
If there are other dtypes that require upcasting, then having a boolean makes sense.

I do like `upcast_to_float64` over `need_upcast`.

Reply:
> If there are other dtypes that require upcasting, then having a boolean makes sense.

Just to make it explicit: at the moment, there's only float32 but other dtype might be supported in the future (though it's not planned). Is this OK for you?

Reply:
> Just to make it explicit: at the moment, there's only float32 but other dtype might be supported in the future (though it's not planned). Is this OK for you?

Yup. That's okay with me.

Reply:
Done via 2a826997bd21f2b2b8bcfaa92e163335555e5afa.